sygnal/helper/proxy/proxy_asyncio.py

```diff
@@ -0,0 +1,321 @@
+        try:
+            response_header, dangling_bytes = self.buffer.split(b"\r\n\r\n", maxsplit=1)
+            lines = response_header.split(b"\r\n")
+            status_line = lines[0]
+            # maxsplit=2 denotes the number of separators, not the № items
+            # StatusLine ← HTTPVersion SP StatusCode SP ReasonPhrase
+            # None of the fields may contain CRLF, and only ReasonPhrase may
+            # contain SP.
+            [http_version, status, reason_phrase] = status_line.split(b" ", maxsplit=2)
+            logger.debug(
+                "CONNECT response from proxy: hv=%s, r=%s, rp=%s",
+                http_version,
+                status,
+                reason_phrase,
+            )
+            if status != b"200":
+                # 200 Successful (aka Connection Established) is what we want
+                # if it is not what we have, then we don't have a tunnel
+                logger.error(
+                    "Error from HTTP Proxy"
+                    " whilst attempting CONNECT: %s (%s);"
+                    "aborting connection.",
+                    status,
+                    reason_phrase,
+                )
+                self.transport.close()
+                raise ProxyConnectError(
+                    "Error from HTTP Proxy"
+                    f" whilst attempting CONNECT: {status.decode()}"
+                    f" ({reason_phrase.decode()}); aborting connection."
+                )
+
+            logger.debug("Ready to switch over protocol")
+
+            self.buffer = None  # type: ignore
+            # TLS doesn't seem to allow the server to talk before the client begins
+            # the handshake, but plain HTTP/2 seems like the server can talk first
+            # and who knows what the future holds?
+            # (we may wish to use other protocols or TLS might change)
+            # So we must also keep the left-over bytes to hand to the next Protocol
+            self._wait_for_establishment.set_result(dangling_bytes)
+        except Exception as exc:
+            logger.error("HTTP CONNECT failed.", exc_info=True)
+            self._wait_for_establishment.set_exception(exc)
+
+    def eof_received(self) -> Optional[bool]:
+        return super().eof_received()
+
+    def connection_made(self, transport: BaseTransport) -> None:
+        if not isinstance(transport, Transport):
+            raise ValueError("transport must be a proper Transport")
+
+        super().connection_made(transport)
+        # when we get a TCP connection to the HTTP proxy, we invoke the CONNECT
+        # method on it to open a tunnelled TCP connection through the proxy to
+        # the other side
+        host, port = self.target_hostport
+        transport.write(f"CONNECT {host}:{port} HTTP/1.1\r\n".encode())
+        parts = self.proxy_url_parts
+        transport.write(f"Host: {parts.hostname}:{parts.port or 80}\r\n".encode())
```

Comment:
wrong default port. Also, I'm pretty sure that convention is that if it's the default port (ie, 443 or 80, depending on your protocol) then you omit the port from the Host header.

Also... it looks like the twisted impl doesn't bother with this? if it's not needed that would simplify the interface.

Reply:
There is no requirement to omit the port if it is default. It may be cleaner but not necessary; that said, should I bother anyway?

However, in my experience netcatting HTTP servers, they choke (400?) if you leave off the Host header... [This is backed up by MDN.](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host)

Reply:
http servers yes; for CONNECT proxies I think the case is a bit different.

anyway if you want to leave it as is, that's fine.

Reply:
Interesting that twisted.web makes special effort to skip the default port: https://github.com/twisted/twisted/blob/d47dafe0717b3877d5d0017ab0feb0721cb78613/src/twisted/web/http.py#L1468-L1476

I kinda think it would be best to go with the flow here and omit it for the default port.

Reply:
I notice that Firefox will set the `Host:` to the same as the `CONNECT` destination.
I can't find any results suggesting that Host is optional for HTTP proxies...
The code nicked from Synapse uses HTTP/1.0 and Host isn't required for HTTP/1.0.

So I'm inclined to go with the flow and switch over to HTTP/1.0 and take away the Host header, in case it has a different meaning to what I expected and causes someone's proxy to get upset.